wordpress-mobile/WordPress-iOS
# Issues/4093 reader list refresh

Refs #4093 

This PR adds the refreshed reader post list to the app, but does not yet replace the current reader post list pending additional work. 

The new `ReaderStreamViewController` ports the current `ReaderPostsViewController` to Swift. Other changes include:
- Adds header views for site, tag, and some list topics. 
- Consolidates the functionality of the (now deprecated) `ReaderBrowseSiteViewController` 
- Adopts storyboards an xibs for layout
- Improvements to the refresh mechanism. 
- Convenience methods for identifying a topic type.  These are subject to change when refreshing the topic model
- Allows sharing posts from the list
- Updates the (now deprecated) `ReaderPostsViewController` to reconcile other changes made to the `ReaderPostService` and `WPTableViewHandler`.

#### Please note: 
- Card layout is pending a refresh to be full screen on the iPhone. 
- Some features in the new header views (e.g. following, some data fields) are not implemented pending improvements to the `ReaderTopic` model in a later.

Of particular interest is this PR changes the `NSManagedObjectContext` used for fetching and displaying content in the new list, and in the old list, to be a child context of our main context, but also running on the main queue.  Syncing is still performed on a private queue, derived context. This is done for better separation of concerns and encapsulation of the mechanism to refresh the table view while preserving the visible content.  cc @astralbodies


#### Things to test:

#### General:
- Does the new list load?
- Does pull to refresh function?
- Does load more function? 

#### Headers: 
- Freshly Pressed, Posts I Like, and Following feeds should not show a header. 
- Tags should show a tag header (following is not implemented)
- Browsing a site should show a site header. Some content is a placeholder. Following is not enabled. 
- Custom lists should show a list header.  Some content is a placeholder.

#### Cell Menus:
- You should not see an option to block a site while browsing a site, or browsing Following.
- You should be able to block a site, and undo a block.  Pull to refresh should clear an undo cell. 
- You should be able to share a post via the share menu option. 

#### Cell Features: 
- You should be able to like or unlike a post by tapping the like button in a cell
- You should be able to view comments by tapping on the comment button in a cell
- You should be able to view a post detail by tapping on a cell

#### Site Browsing:
- Tapping the site name or blavatar should browse a site. 
- When browsing a site, the site name should be disabled. Tapping the site name or blavatar should be like any other tap on a cell. 

#### Old list:
Pull to refresh should still work. 
Infinite scroll / load more should still work. 
You should still be able to like or unlike a post from a cell

### How to test
You'll need to edit `ReaderViewController.m` to use `ReaderStreamViewController` instead of `ReaderPostViewController`.  The relevant changes are in [this gist](https://gist.github.com/aerych/feeff93b445adb7d655c).

@jleandroperez, can I trouble you for a review?  I expect there are a few opportunities to improve my Swift code that you could suggest.

Needs Review: @jleandroperez
